kitsune/community/api.py

```diff
@@ -0,0 +1,330 @@
+from kitsune.sumo.api import GenericAPIException
+from kitsune.users.models import UserMappingType
+from kitsune.wiki.models import RevisionMetricsMappingType
+
+
+# This should be the higher than the max number of contributors for a
+# section.  There isn't a way to tell ES to just return everything.
+BIG_NUMBER = 10000
+
+
+class TopContributorsBase(views.APIView):
+
+    def get(self, request):
+        return Response(self.get_data(request))
+
+    def get_filters(self):
+        self.filter_values = self.get_default_filters()
+        # request.GET is a multidict, so simple `.update(request.GET)` causes
+        # everything to be a list. This converts it into a plain single dict.
+        self.filter_values.update(dict(self.request.GET.items()))
+
+        f = F()
+
+        for key, value in self.filter_values.items():
+            method_name = 'filter_' + key
+            if not hasattr(self, method_name):
+                raise GenericAPIException(400, 'Unknown filter field {}'.format(key))
+            filter_method = getattr(self, 'filter_' + key)
+            f &= filter_method(value)
+
+        return f
+
+    def get_default_filters(self):
+        return {
+            'startdate': (datetime.now() - timedelta(days=90)).strftime('%Y-%m-%d'),
+            'enddate': datetime.now().strftime('%Y-%m-%d'),
+        }
+
+    def filter_page(self, value):
+        """No-op, so that unknown filters can be blocked."""
+        return F()
+
+    def filter_ordering(self, value):
+        """No-op, so that unknown filters can be blocked."""
+        return F()
+
+    def filter_startdate(self, value):
+        date = fields.DateField().from_native(value)
+        dt = datetime.combine(date, datetime.min.time())
+        return F(created__gte=dt)
+
+    def filter_enddate(self, value):
+        date = fields.DateField().from_native(value)
+        dt = datetime.combine(date, datetime.max.time())
+        return F(created__lte=dt)
+
+    def filter_username(self, value):
+        username_lower = value.lower()
+
+        username_filter = (
+            F(iusername__prefix=username_lower) |
+            F(idisplay_name__prefix=username_lower) |
+            F(itwitter_usernames__prefix=username_lower))
+
+        users = UserMappingType.reshape(
+            UserMappingType
+            .search()
+            .filter(username_filter)
+            .values_dict('id')
+            [:BIG_NUMBER])
+
+        return F(creator_id__in=[u['id'] for u in users])
+
+    def filter_locale(self, value):
+        return F(locale=value)
+
+
+class TopContributorsQuestions(TopContributorsBase):
+
+    def get_default_filters(self):
+        filters = super(TopContributorsQuestions, self).get_default_filters()
+        filters['ordering'] = '-answer_count'
+        return filters
+
+    def get_filters(self):
+        f = super(TopContributorsQuestions, self).get_filters()
+        f &= F(by_asker=False)
```

Comment:
These are probably named poorly. `get_default_filters` is the default values for things users of the API may set. Users can't change this aspect of the API, it is just always true.